donkeycar/pipeline/types.py

```diff
@@ -70,22 +71,50 @@ class TubDataset(object):
     '''
 
     def __init__(self, config: Config, tub_paths: List[str],
-                 shuffle: bool = True) -> None:
+                 shuffle: bool = True, select: Optional[str] = None) -> None:
         self.config = config
         self.tub_paths = tub_paths
         self.shuffle = shuffle
+        self.select = None
         self.tubs: List[Tub] = [Tub(tub_path, read_only=True)
                                 for tub_path in self.tub_paths]
         self.records: List[TubRecord] = list()
+        if select:
+            # add '_index' to the allowed field list as this is
+            # auto-generated by the tub and not part of the inputs
+            fields = self.tubs[0].manifest.inputs + ['_index']
+            self.select = create_filter_string(select, fields)
 
     def train_test_split(self) -> Tuple[List[TubRecord], List[TubRecord]]:
-        print(f'Loading tubs from paths {self.tub_paths}')
+        msg = f'Loading tubs from paths {self.tub_paths}' + f' with filter ' \
+              f'{self.select}' if self.select else ''
+        print(msg)
         self.records.clear()
         for tub in self.tubs:
             for underlying in tub:
-                record = TubRecord(self.config, tub.base_path,
-                                   underlying=underlying)
-                self.records.append(record)
+                record = TubRecord(self.config, tub.base_path, underlying)
+                # eval string uses 'record' as variable so this parses
+                if not self.select or eval(self.select):
```

Comment:
We should not be using `eval` for anything. Its extremely unsafe.

Reply:
I haven't found an alternative that can parse expressions like `user/throttle > 0.1 and _index < 1000`. Apart from that, it is not used anywhere in the running vehicle, only in the selection of data for training, and it is non-destructive, it just omits records. You can test the expression in the UI, it shows you if it evaluates to true or false. It's neither put into a public web server - that wouldn't be a good idea, nor using an unchecked string as input. I don't think there is any more safety concern for malfunction than opening a python prompt on your machine and running the same code there in case anyone would put `shutil.rmtree('/')`.

Reply:
I think if we switch to the actual lambda in the `config` then we don't need to do anything. (That is what i meant)

Reply:
I've switched to that in training.